Response Caching Middleware topic updates for 2.0

guardrex opened the pull request:
Fixes #4019 

[Review site](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/middleware?branch=pr-en-us-4020&tabs=aspnetcore2x)

dnfclas commented:
@guardrex,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

guardrex commented:
@Rick-Anderson @tdykstra @JunTaoLuo 

Notes:
* Adopted the sample layout spec with topic tabs (with 2.x tab first)
* Updated the 1.x sample to latest package versions on 1.x.y.
* 😵 Nutjob:tm: 😵 startup pattern (i.e., using extension methods of `WebHost` for startup in 2.0 sample ... one of my fav 2.0 features)
* Cleaned up the text; the usual stuff ...
   - Addressed guidelines adherence issues (contractions, :hocho: the future tense, etc.)
   - "application" :point_right: "app"
   - Better word and phrase choices
* 🥀 Table Beautification 🌷 (and **\<p>** wasn't rendering well in table cells, so I've removed those from the tables ... all good now)
* Added technology labeling "ASP.NET Core" to title and **\<h1>**
* Updated for new features of 2.0, but I definitely need John's :eye: on those bits.

*It's off of WIP and ready for* :eyes:!

I'm moving on to 🏰 *Ye Olde Response Compression Middleware* 🏰 topic next.

Rick-Anderson commented on aspnetcore/performance/caching/middleware.md:
```diff
@@ -1,48 +1,65 @@
 
 ## Package
-To include the middleware in your project, add a reference to the  [`Microsoft.AspNetCore.ResponseCaching`](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCaching/) package. The middleware depends on .NET Framework 4.5.1 or .NET Standard 1.3 or later. This feature is available for apps that target ASP.NET Core 1.1.0 or later.
+To include the middleware in a project, add a reference to the  [`Microsoft.AspNetCore.ResponseCaching`](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCaching/) package. This feature is available for apps that target ASP.NET Core 1.1 or later.
```
Either add tabs or just mention you don't need to add this package in 2.0 with the all metadata package.

Rick-Anderson commented on aspnetcore/performance/caching/middleware.md:
```diff
@@ -1,48 +1,65 @@
->[!WARNING]
+---
+
+Configure the app to use the middleware when processing requests. The sample app adds a [`Cache-Control`](https://tools.ietf.org/html/rfc7234#section-5.2) header to the response that caches cacheable responses for up to 10 seconds. The sample sends a [`Vary`](https://tools.ietf.org/html/rfc7231#section-7.1.4) header to configure the middleware to serve a cached response only if the [`Accept-Encod ...
```
>Configure the app to use the middleware when processing requests.

I don't understand this command out of the blue.

guardrex replied:
> this command

If you mean the `Cache-Control` header bit, I wonder if this should just refer back to the topic we hope they read first it's mentioned in the opening para): https://docs.microsoft.com/en-us/aspnet/core/performance/caching/response. I can add a quick statement to that effect.

Otherwise, let me know which bit needs more work ... the `UseResponseCaching`, the `app.Run`, the `Cache-Control`, or the `WriteAsync` bit, and I'll add more text here.

Rick-Anderson replied:
I don't understand "Configure the app to use the middleware when processing requests"

guardrex replied:
ok ... I'll fix that on the next commit.

Rick-Anderson commented on aspnetcore/performance/caching/middleware.md:
```diff
@@ -1,48 +1,65 @@
+| Option                | Default Value |
+| --------------------- | ------------- |
+| UseCaseSensitivePaths | Determines if responses are cached on case-sensitive paths.</p><p>The default value is `false`. |
+| MaximumBodySize       | The largest cacheable size for the response body in bytes.</p>The default value is `64 * 1024 * 1024` [64 MB (67,108,864 bytes)]. |
```
I think it reads better without the exact bytes.
(67,108,864 bytes)

guardrex replied:
I'm still a bit concerned about dropping it even with the `64 * 1024 * 1024`. I hope to stay as far away as possible from the giant can of slimy worms with MB and MiB. The reason I like spelling it out is that I don't know if it's immediately obvious with `64 * 1024 * 1024` that 64 MB isn't 64,000,000 bytes. By explicitly stating the bytes, there is *no* confusion at all, ever, on this point.

Is this a suggestion or do u prefer it without the explicit bytes. I'd leave it; but of course, I'll remove it if you prefer it out of there.

Rick-Anderson replied:
I'm just recommending dropping The default value is `64 * 1024 * 1024` [64 MB (~67,108,864 bytes~)].

Rick-Anderson commented on aspnetcore/performance/caching/middleware.md:
```diff
@@ -1,48 +1,65 @@
+| --------------------- | ------------- |
+| UseCaseSensitivePaths | Determines if responses are cached on case-sensitive paths.</p><p>The default value is `false`. |
+| MaximumBodySize       | The largest cacheable size for the response body in bytes.</p>The default value is `64 * 1024 * 1024` [64 MB (67,108,864 bytes)]. |
+| SizeLimit             | The size limit for the response cache middleware in bytes. The default value is `100 * 1024 * 1024` [100 MB (104,857,600 bytes)]. |
```
remove exact bytes then you can use (100 MB) rather than [100 MB]

Rick-Anderson commented on aspnetcore/performance/caching/middleware.md:
````diff
@@ -55,9 +72,9 @@ services.AddResponseCaching(options =>
 ```
 
 ## VaryByQueryKeys feature
````
## VaryByQueryKeys ~feature~

Rick-Anderson commented:
>When not using MVC, you can vary response caching with the `VaryByQueryKeys` feature. Use the `ResponseCachingFeature` directly from the `IFeatureCollection` of the `HttpContext`:

@JunTaoLuo  when would you not be using MVC?

Rick-Anderson commented on aspnetcore/performance/caching/middleware.md:
```diff
@@ -68,46 +85,50 @@ if (responseCachingFeature != null)
-If caching behavior is not as you expect, confirm that responses are cacheable and capable of being served from the cache by examining the request's incoming headers and the response's outgoing headers. The conditions by which a response will be cached are listed below.
+If caching behavior isn't as you expect, confirm that responses are cacheable and capable of being served from the cache by examining the request's incoming headers and the response's outgoing headers. The conditions by which a response are cached are listed below.
+
+Enabling logging can help when debugging. The middleware logs caching behavior and when a response is retrieved from cache. See [Logging in ASP.NET Core](xref:fundamentals/logging) for more information on enabling logging.
```
See [Logging in ASP.NET Core](xref:fundamentals/logging) for more information ~on enabling logging.~

Rick-Anderson replied:
Or even
Enabling `[logging](xref:fundamentals/logging)`  can help when debugging. The middleware logs caching behavior and when a response is retrieved from cache. ~See [Logging in ASP.NET Core](xref:fundamentals/logging) for more information on enabling logging.~

guardrex commented:
@Rick-Anderson Done except ....... use "MiB" ??? What is the current MS position on that? I don't have access to the style manual (and kind'a even doubt it's in there).

Rick-Anderson commented on aspnetcore/performance/caching/middleware.md:
```diff
@@ -1,48 +1,65 @@
+| Option                | Default Value |
+| --------------------- | ------------- |
+| UseCaseSensitivePaths | Determines if responses are cached on case-sensitive paths.</p><p>The default value is `false`. |
+| MaximumBodySize       | The largest cacheable size for the response body in bytes.</p>The default value is `64 * 1024 * 1024` [64 MiB]. |
```
Why use `[]`? I'd go with
The default value is `64 * 1024 * 1024` (64 MB).
I don't think that will cause any confusion. MiB is confusing.

guardrex replied:
Yeah ... MiB *is* confusing ... correct ... but confusing. :smile:

Let's roll the 🎲 🎲 on just having the math there. I'll fix it now (and the \[]).

JunTaoLuo commented:
@Rick-Anderson in case where you are writing a blank web app without MVC, then you can still use the `VaryByQueryKeys` feature via the `ResponseCachingFeature`

JunTaoLuo commented on aspnetcore/performance/caching/middleware/samples/2.x/ResponseCachingSample.csproj:
```diff
@@ -0,0 +1,13 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore" Version="2.0.0" />
```
Curious why we're not using `Microsoft.AspNetCore.All` since that's what is used in the default templates?

guardrex replied:
Unfriendly to non-Core apps. This way, it cross-compiles without conditions in the project file.

Rick-Anderson replied:
@guardrex  - arg missed that. Yes, we use ALL for ALL 2.0 projects

guardrex replied:
Then ur conditioning the use of the package?

Rick-Anderson replied:
We don't care about non-core apps

Rick-Anderson replied:
It's not that we don't care, it's that we want to promote Core and the ALL metapackage

guardrex replied:
Meany!!!! lol ... actually, I slightly misspoke ... I meant non-`netcoreapp` .NET Core apps.

guardrex replied:
This cross-compiles `netXY` projects without having to condition the `All` package.

davidfowl replied:
Show the all package in all samples.

guardrex replied:
kk ... I'll fix it and patch the others after master goes live.

Rick-Anderson commented:
>@Rick-Anderson in case where you are writing a blank web app without MVC, then you can still use the VaryByQueryKeys feature via the ResponseCachingFeature

That's so niche that anyone doing that wouldn't need the info. For the rest of the folks it makes that section confusing.  
@JunTaoLuo  do you agree or should we leave that? I could really simply the paragraph before that if we could remove that trivia statement.

JunTaoLuo commented on aspnetcore/performance/caching/middleware.md:
```diff
@@ -1,48 +1,65 @@
 
 ## Package
-To include the middleware in your project, add a reference to the  [`Microsoft.AspNetCore.ResponseCaching`](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCaching/) package. The middleware depends on .NET Framework 4.5.1 or .NET Standard 1.3 or later. This feature is available for apps that target ASP.NET Core 1.1.0 or later.
+To include the middleware in a project, add a reference to the [`Microsoft.AspNetCore.ResponseCaching`](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCaching/) package or use the [`Microsoft.AspNetCore.All`](https://www.nuget.org/packages/Microsoft.AspNetCore.All/) package. This feature is available for apps that target ASP.NET Core 1.1 or later.
```
Actually it's available in ASP.NET Core 1.0 as well.

guardrex replied:
I've marked this **WIP** for a sec while I make the updates ... here and to the 2.0 sample.

JunTaoLuo commented:
@Rick-Anderson I would prefer leaving that section in since this is the only place that is documented. Otherwise there's no way to discover this feature. It is niche but maybe we can rework the language instead of taking it out completely?

guardrex commented:
@Rick-Anderson U prob don't need to sweat rolling back the other PR's for sample updates.

My turnaround time ain't that long! :grinning:

I'll attend to them instantly (after they go live); tonight or tomorrow morning.

guardrex commented:
@Rick-Anderson @davidfowl One point of clarification tho: You *do* or *don't* want samples targeting netfx? I don't need to condition the `All` package if you just want the 2.0 samples to target `netcoreapp2.0`.

If we're going to target `net462` and thus condition the `All` package, do you want the condition on the package reference or do you want it on the item group (even if the `All` package is the only thing in the group)?

davidfowl commented:
> @Rick-Anderson @davidfowl One point of clarification tho: You do or don't want samples targeting netfx? I don't need to condition the All package if you just want the 2.0 samples to target netcoreapp2.0.

I would optimize for .NET Core and have a general document about the differences between ASP.NET Core on .NET Framework.

guardrex commented:
@Rick-Anderson See last commit for the updates.

Rick-Anderson commented:
@guardrex  is this ready to merge?

guardrex commented:
@Rick-Anderson Yes, ready for :rocket:!
